core/vm/environment.libevm.go

```diff
@@ -134,13 +152,17 @@ func (e *environment) callContract(typ CallType, addr common.Address, input []by
 
 		startGas := gas
 		defer func() {
-			t.CaptureEnd(retData, startGas-retGas, retErr)
+			t.CaptureEnd(retData, startGas-e.Gas(), retErr)
 		}()
 	}
 
 	switch typ {
 	case Call:
-		return e.evm.Call(caller, addr, input, gas, value)
+		ret, returnGas, err := e.evm.Call(caller, addr, input, gas, value)
+		if err := e.refundGas(returnGas); err != nil {
+			return nil, err
+		}
+		return ret, err
```

Comment:
nit err is nil here, so we could for simplicity return just

```suggestion
		return ret, nil
```

Reply:
I've renamed it to `callErr` to make it clearer that it's actually the error from `e.evm.Call()`, which may not be `nil` here. This part of the EVM can return both `ret` and `err` as non-nil (and non-empty for the bytes) in the case of a transaction reverting and it's important that both are propagated.

Reply:
Oh whoa nifty, I did not see that! Thanks for clarifying!